src/app/jscaip/MGPNode.ts

```diff
@@ -1,268 +1,185 @@
-            currentPlayer === Player.ZERO ?
-                ((a: number, b: number): boolean => a < b) :
-                ((a: number, b: number): boolean => b < a);
-        for (const move of possibleMoves) {
-            const child: MGPNode<R, M, S, L, B> = this.getOrCreateChild(move, minimax);
-            const bestChildDescendant: MGPNode<R, M, S, L, B> =
-                child.alphaBeta(depth - 1, alpha, beta, minimax, random, prune);
-            const bestChildValue: number = bestChildDescendant.getHopedValue(minimax);
-            if (newValueIsBetter(bestChildValue, extremumExpected) || bestChilds.length === 0) {
-                extremumExpected = bestChildValue;
-                bestChilds = [bestChildDescendant];
-            } else if (bestChildValue === extremumExpected) {
-                bestChilds.push(bestChildDescendant);
-            }
-            if (prune && newValueIsBetter(extremumExpected, currentPlayer === Player.ZERO ? alpha : beta)) {
-                // cut-off, no need to explore the other childs
-                break;
-            }
-            if (currentPlayer === Player.ZERO) {
-                beta = Math.min(extremumExpected, beta);
-            } else {
-                alpha = Math.max(extremumExpected, alpha);
-            }
-        }
-        return bestChilds;
+    /**
+     * Checks whether this node has children
+     */
+    public hasChildren(): boolean {
+        return this.getChildren().length > 0;
     }
-    private getBestChildAmongst(bestChilds: MGPNode<R, M, S, L, B>[], random: boolean): MGPNode<R, M, S, L, B> {
-        if (random) {
-            const randomIndex: number = Math.floor(Math.random() * bestChilds.length);
-            return bestChilds[randomIndex];
-        } else {
-            return bestChilds[0];
-        }
+    /**
+     * Returns all the children of the node
+     */
+    public getChildren(): GameNode<M, S>[] {
+        return this.children.listValues();
     }
-    private getOrCreateChild(move: M, minimax: Minimax<M, S, L, B>): MGPNode<R, M, S, L, B> {
-        let child: MGPOptional<MGPNode<R, M, S, L, B>> = this.getSonByMove(move);
-        if (child.isAbsent()) {
-            const legality: MGPFallible<L> = minimax.ruler.isLegal(move, this.gameState);
-            const moveString: string = move.toString();
-            assert(legality.isSuccess(), 'The minimax "' + minimax.name + '" has proposed an illegal move (' + moveString + '), refusal reason: ' + legality.getReasonOr('') + ' this should not happen.');
-            const state: S = minimax.ruler.applyLegalMove(move, this.gameState, legality.get());
-            child = MGPOptional.of(new MGPNode(state, MGPOptional.of(this), MGPOptional.of(move), minimax));
-            this.childs.get().push(child.get());
-        }
-        return child.get();
+    /**
+     * Adds a child to this node.
+     */
+    public addChild(move: M, node: GameNode<M, S>): void {
```

Comment:
any check that node has indeed "previousMove" set to 'move' ? Or don't use 'move' cause it's already in the node and ensures this !

Reply:
bien vu, je supprime l'argument move